lib/shared/s3-packages.js

```diff
@@ -0,0 +1,216 @@
+
+  return request.getAsync(bucketUrl)
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    .get('body')
+    .then(xml.parseStringAsync)
+    .get('ListBucketResult')
+    .then((bucketResult) => {
+      return _.get(bucketResult, [ 'Contents' ], []);
+    })
+    .reduce((accumulator, entry) => {
+      const [ name, version ] = _.split(_.first(entry.Key), '/');
+
+      if (name === S3_PACKAGE_NAME) {
+        if (_.isNil(accumulator[version])) {
+          accumulator[version] = 1;
+        } else {
+          accumulator[version] += 1;
+        }
+      }
+
+      return accumulator;
+    }, [])
+    .then((versions) => {
+      return _.keys(_.pickBy(versions, (occurrences) => {
+        return occurrences >= NUMBER_OF_PACKAGES;
+      }));
+    })
+    .catch({
+      code: 'ENOTFOUND'
+    }, {
+      code: 'ETIMEDOUT'
+    }, () => {
+      return [];
+    });
+});
+
+/**
+ * @summary Get the latest available version for a given release type
+ * @function
+ * @public
+ *
+ * @param {String} releaseType - release type
+ * @param {Object} options - options
+ * @param {String} [options.range] - semver range
+ * @fulfil {(String|Undefined)} - latest version
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getLatestVersion(release.RELEASE_TYPE.PRODUCTION, {
+ *   range: '>=2.0.0'
+ * }).then((latestVersion) => {
+ *   console.log(`The latest version is: ${latestVersion}`);
+ * });
+ */
+exports.getLatestVersion = (releaseType, options = {}) => {
+
+  // For manual testing purposes
+  const ETCHER_FAKE_S3_LATEST_VERSION = process.env.ETCHER_FAKE_S3_LATEST_VERSION;
+  if (!_.isNil(ETCHER_FAKE_S3_LATEST_VERSION)) {
+    if (release.getReleaseType(ETCHER_FAKE_S3_LATEST_VERSION) === releaseType) {
+      return Bluebird.resolve(ETCHER_FAKE_S3_LATEST_VERSION);
+    }
+
+    return Bluebird.resolve();
+  }
+
+  const bucketUrl = exports.getBucketUrlFromReleaseType(releaseType);
+  if (_.isNil(bucketUrl)) {
+    return Bluebird.reject(new Error(`No bucket URL found for release type: ${releaseType}`));
+  }
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return exports.getRemoteVersions(bucketUrl).filter((version) => {
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    // This check allows us to ignore snapshot builds in production
+    // buckets, and viceversa, which could have been uploaded by mistake.
+    const versionReleaseType = release.getReleaseType(version);
+    if (versionReleaseType !== releaseType) {
+      return false;
+    }
+
+    // The `semver` module refuses to apply ranges to prerelease versions
+    // As a workaround, we drop the prerelease tags, if any, apply the range
+    // on that, and keep using the prerelease tag from then on.
+    // See https://github.com/npm/node-semver#prerelease-tags
+    const strippedVersion = `${semver.major(version)}.${semver.minor(version)}.${semver.patch(version)}`;
+
+    return semver.satisfies(strippedVersion, options.range || '*');
```

Comment:
As @pimterry sorta hinted at, I wonder if it'd be worth extracting out our own custom "semver satisfies" function (that _does_ cope with prerelease tags), and then use that here instead of dealing with the `strippedVersion` ourselves directly inside `getRemoteVersions` ?